api/src/main/java/io/minio/MinioClient.java

```diff
@@ -5762,6 +5763,131 @@ public void listenBucketNotification(String bucketName, String prefix, String su
+   *
+   */
+  public CloseableIterator<Result<NotificationInfo>> listenBucketNotification(String bucketName, String prefix,
+      String suffix, String[] events)
+    throws IOException, InvalidKeyException, NoSuchAlgorithmException, InsufficientDataException,
+    InvalidResponseException, InternalException, NoResponseException, InvalidBucketNameException,
+    XmlPullParserException, ErrorResponseException {
+
+    Multimap<String, String> queryParamMap = HashMultimap.create();
+    queryParamMap.put("prefix", prefix);
+    queryParamMap.put("suffix", suffix);
+    for (String event : events) {
+      queryParamMap.put("events", event);
+    }
+
+    HttpResponse response = executeReq(Method.GET, getRegion(bucketName),
+        bucketName, "", null, queryParamMap, null, 0);
+
+    return new CloseableIterator<Result<NotificationInfo>>() {
+      Scanner scanner  = new Scanner(response.body().charStream()).useDelimiter("\n");
+
+      String notificationString = null;
+      ObjectMapper mapper = new ObjectMapper();
+      NotificationInfo notificationInfo = null;
+      boolean isClosed = false;
+
+      @Override
+      public void close() throws IOException {
+        if (!isClosed) {
+          try {
+            response.body().close();
+            scanner.close();
+          } finally {
+            isClosed = true;
+          }
+        }
+      }
+
+      public boolean populate()  {
+        if (isClosed) {
+          return false;
+        }
+
+        if (notificationString != null) {
+          return true;
+        }
+
+        while (scanner.hasNext()) {
+          notificationString = scanner.next().trim();
+          if ( !notificationString.equals("")) {
+            break;
+          }
+        }
+
+        if (notificationString  == null || notificationString.equals("")) {
+          try {
+            close();
+          } catch (IOException e) {
+            isClosed = true;
+          }
+          return false;
+        }
+        return true;
+      }
+
+      @Override
+      public boolean hasNext() {
+        return populate();
+      }
+
+      @Override
+      public Result<NotificationInfo> next() {
+        if (isClosed) {
+          throw new NoSuchElementException();
+        }
+        if ((notificationString  == null || notificationString.equals("")) &&  !populate() ) {
+              throw  new NoSuchElementException();
+        }
+
+        try {
+              notificationInfo = mapper.readValue(notificationString, NotificationInfo.class);
+              return new Result<>(notificationInfo, null);
+            } catch ( IOException ex ) {
+              if ( ex instanceof JsonParseException) {
+                return new Result<>(null, (JsonParseException)ex);
+              } else if ( ex instanceof  JsonMappingException) {
+                return new Result<>(null, (JsonMappingException)ex);
+              } else {
+                return new Result<>(null, ex);
```

Comment:
below code is correct.

```java
} catch (JsonParseException | JsonMappingException | IOException e) {
    return new Result(null, e);
}
```
In `Result.java`, you need to have `get()` like below (Note: I modified `if` block better)
```java
  public T get()
    throws InvalidBucketNameException, NoSuchAlgorithmException, InsufficientDataException, IOException,
           InvalidKeyException, NoResponseException, XmlPullParserException, ErrorResponseException,
           JsonParseException, JsonMappingException, InternalException {
    if (ex == null) {
      return type;
    }

    if (ex instanceof InvalidBucketNameException) {
      throw (InvalidBucketNameException) ex;
    }

    if (ex instanceof NoSuchAlgorithmException) {
      throw (NoSuchAlgorithmException) ex;
    }

    if (ex instanceof InsufficientDataException) {
      throw (InsufficientDataException) ex;
    }

    if (ex instanceof IOException) {
      throw (IOException) ex;
    }

    if (ex instanceof InvalidKeyException) {
      throw (InvalidKeyException) ex;
    }

    if (ex instanceof NoResponseException) {
      throw (NoResponseException) ex;
    }

    if (ex instanceof XmlPullParserException) {
      throw (XmlPullParserException) ex;
    }

    if (ex instanceof ErrorResponseException) {
      throw (ErrorResponseException) ex;
    }

    if (ex instanceof JsonParseException) {
      throw (JsonParseException) ex;
    }

    if (ex instanceof JsonMappingException) {
      throw (JsonMappingException) ex;
    }

    throw (InternalException) ex;
  }
}
```

Reply:
Since `JsonParseException ` and `JsonMappingException` are sub classes of `IOException` , the pipe in catch would not work. So i have caught the child class exception first and then the parent class.
Refer https://stackoverflow.com/questions/11116853/declare-method-to-throw-an-exception-and-subclass-of-this-exception

```
            } catch (JsonParseException e) {
              return new Result<>(null, e);
            } catch ( JsonMappingException e) {
              return new Result<>(null, e);
            } catch ( IOException e) {
              return new Result<>(null, e);
```

Reply:
:+1: You would need to do the same in `Result.get()`

Reply:
@sinhaashish Is the change requested in `Result.get()` done?

Reply:
@balamurugana  this code is already added in Result.java.
```

    if (ex instanceof JsonParseException) {
      throw (JsonParseException) ex;
    }

    if (ex instanceof JsonMappingException) {
      throw (JsonMappingException) ex;
    }

    if (ex instanceof IOException) {
      throw (IOException) ex;
    }
```